source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1739 @@
+      // the FastScape cell size can be properly set.
+      std::array<unsigned int,dim> repetitions = geometry->get_repetitions();
+
+      // Set number of x points, which is generally 1+(FastScape refinement level)^2.
+      // The FastScape refinement level is a combination of the maximum ASPECT refinement level
+      // at the surface and any additional refinement we want in FastScape. If
+      // repetitions are specified we need to adjust the number of points to match what ASPECT has,
+      // which can be determined by multiplying the points by the repetitions before adding 1.
+      // Finally, if ghost nodes are used we add two additional points on each side.
+      //nx = 1+(2*use_ghost_nodes)+std::pow(2,maximum_surface_refinement_level+additional_refinement)*repetitions[0];
+      const unsigned int ghost_nodes = 2*use_ghost_nodes;
+      const unsigned int fastscape_refinement_level = maximum_surface_refinement_level + additional_refinement;
+      const unsigned int fastscape_nodes = std::pow(2,fastscape_refinement_level);
+      nx = fastscape_nodes * repetitions[0] + ghost_nodes + 1;
+
+      // Size of FastScape cell.
+      dx = (grid_extent[0].second)/(fastscape_nodes * repetitions[0]);
+
+      // FastScape X extent, which is generally ASPECT's extent unless the ghost nodes are used,
+      // in which case 2 cells are added on either side.
+      x_extent = (grid_extent[0].second) + dx * ghost_nodes;
+
+      // Sub intervals are 3 less than points, if including the ghost nodes. Otherwise 1 less.
+      table_intervals[0] = fastscape_nodes * repetitions[0];
+      table_intervals[dim-1] = 1;
+
+      if (dim == 2)
+        {
+          dy = dx;
+          y_extent = round(y_extent_2d/dy)*dy + dy * ghost_nodes;
+          ny = 1+y_extent/dy;
+        }
+      else
+        {
+          ny = fastscape_nodes * repetitions[1] + ghost_nodes + 1;
+          dy = (grid_extent[1].second)/(fastscape_nodes * repetitions[1]);
+          table_intervals[1] = fastscape_nodes * repetitions[1];
+          y_extent = (grid_extent[1].second) + dy * ghost_nodes;
+        }
+
+      // Determine array size to send to FastScape
+      array_size = nx*ny;
+
+      // Create a folder for the FastScape visualization files.
+      Utilities::create_directory (this->get_output_directory() + "fastscape/",
+                                   this->get_mpi_communicator(),
+                                   false);
+
+      last_output_time = 0;
+    }
+
+
+    template <int dim>
+    void
+    FastScape<dim>::compute_velocity_constraints_on_boundary(const DoFHandler<dim> &mesh_deformation_dof_handler,
+                                                             AffineConstraints<double> &mesh_velocity_constraints,
+                                                             const std::set<types::boundary_id> &boundary_ids) const
+    {
+      if (this->get_timestep_number() == 0)
+        return;
```

Comment:
Yes, if I remember correctly it was something like this. It looks like the mesh displacement system is one of the first things called, even before adding initial topography. Also, since there is no timestep length to yet run fastscape I figured it would be easiest to initialize and run it on timestep 1.

Reply:
I see. Would you mind encoding your understanding of why you skip the first time step in a comment above these two lines?